In [1]:
!pip install -q -U google-generativeai

In [2]:
import google.generativeai as genai 

/home/manish/anaconda3/envs/cuda121/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [ ]:
google_api = 'AIzaSyAMlYqwvuQgekl8nlqc56XTqJVBufszrBU'
#  AIzaSyAMlYqwvuQgekl8nlqc56XTqJVBufszrBU
genai.configure(api_key=google_api)

In [5]:
for m in genai.list_models():
    if 'generateContent' in m.supported_generation_methods:
        print(m.name)

models/gemini-1.5-pro-latest
models/gemini-1.5-pro-002
models/gemini-1.5-pro
models/gemini-1.5-flash-latest
models/gemini-1.5-flash
models/gemini-1.5-flash-002
models/gemini-1.5-flash-8b
models/gemini-1.5-flash-8b-001
models/gemini-1.5-flash-8b-latest
models/gemini-2.5-pro-preview-03-25
models/gemini-2.5-flash-preview-05-20
models/gemini-2.5-flash
models/gemini-2.5-flash-lite-preview-06-17
models/gemini-2.5-pro-preview-05-06
models/gemini-2.5-pro-preview-06-05
models/gemini-2.5-pro
models/gemini-2.0-flash-exp
models/gemini-2.0-flash
models/gemini-2.0-flash-001
models/gemini-2.0-flash-exp-image-generation
models/gemini-2.0-flash-lite-001
models/gemini-2.0-flash-lite
models/gemini-2.0-flash-preview-image-generation
models/gemini-2.0-flash-lite-preview-02-05
models/gemini-2.0-flash-lite-preview
models/gemini-2.0-pro-exp
models/gemini-2.0-pro-exp-02-05
models/gemini-exp-1206
models/gemini-2.0-flash-thinking-exp-01-21
models/gemini-2.0-flash-thinking-exp
models/gemini-2.0-flash-thinking-exp

In [6]:

# Model Configuration
MODEL_CONFIG = {
  "temperature": 0.2,
  "top_p": 1,
  "top_k": 32,
  "max_output_tokens": 4096,
}

## Safety Settings of Model
safety_settings = [
  {
    "category": "HARM_CATEGORY_HARASSMENT",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  },
  {
    "category": "HARM_CATEGORY_HATE_SPEECH",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  },
  {
    "category": "HARM_CATEGORY_SEXUALLY_EXPLICIT",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  },
  {
    "category": "HARM_CATEGORY_DANGEROUS_CONTENT",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  }
]

In [59]:
model = genai.GenerativeModel(
                            model_name='gemini-2.5-flash',
                            # model_name='gemini-2.5-pro',
                              generation_config=MODEL_CONFIG,
                              safety_settings=safety_settings)
model

genai.GenerativeModel(
    model_name='models/gemini-2.5-flash',
    generation_config={'temperature': 0.2, 'top_p': 1, 'top_k': 32, 'max_output_tokens': 4096},
    safety_settings={<HarmCategory.HARM_CATEGORY_HARASSMENT: 7>: <HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE: 2>, <HarmCategory.HARM_CATEGORY_HATE_SPEECH: 8>: <HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE: 2>, <HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: 9>: <HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE: 2>, <HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: 10>: <HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE: 2>},
    tools=None,
    system_instruction=None,
    cached_content=None
)

In [60]:
from pathlib import Path

def image_format(image_path):
    img = Path(image_path)

    if not img.exists():
        raise FileNotFoundError(f"Could not find image: {img}")

    image_parts = [
        {
            "mime_type": "image/png", ## Mime type are PNG - image/png. JPEG - image/jpeg. WEBP - image/webp
            "data": img.read_bytes()
        }
    ]
    return image_parts


image_format('/home/manish/Desktop/projects/etfp/bill_image_receipt.png')

[{'mime_type': 'image/png',
  'data': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x02\x83\x00\x00\x04M\x08\x02\x00\x00\x00\xa6>\xdd\'\x00\x00J\xaeIDATx\xda\xed\xdd\xe1\x92\xa3:\xb2\xa8\xd1~\xff\x97\xe6F\x9c\x1b11\xd3]\x05\xa9\xcc\x94\x10\xb0\xd6\xaf\x99\xde.\x8c1\xe8\x03\x8c\xf1\x9f\x03\x00\xb8\xcf\x1f\x8b\x00\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\xd9xu\xa99\x9f\xda\xb3^\xbb\x95a\xea\x9a\xe3\xbd@\x89A\x89\x8d\xfe\xf7\xaf9\xde\x0b\x94\x18\x94\xd8\xe8\x7f\xf3\x9a\xe3\xbd@\x89A\x89\x8d\xfe7\xaf9\xde\x0b\x94\x

In [61]:
def gemini_output(image_path,
                  system_prompt,
                  user_prompt):
    
    image_info = image_format(image_path)
    print(image_info[0])
    input_prompt = [system_prompt, image_info[0], user_prompt]
    response = model.generate_content(input_prompt)

    return response.text




In [62]:
system_prompt = """
               You are a specialist in comprehending receipts.
               Input images in the form of receipts will be provided to you,
               and your task is to respond to questions based on the content of the input image.
               """

image_path = "/home/manish/Desktop/projects/etfp/bill_image_receipt.png"

user_prompt = "What is the balance amount in the image?"

gemini_output(image_path, system_prompt, user_prompt)

{'mime_type': 'image/png', 'data': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x02\x83\x00\x00\x04M\x08\x02\x00\x00\x00\xa6>\xdd\'\x00\x00J\xaeIDATx\xda\xed\xdd\xe1\x92\xa3:\xb2\xa8\xd1~\xff\x97\xe6F\x9c\x1b11\xd3]\x05\xa9\xcc\x94\x10\xb0\xd6\xaf\x99\xde.\x8c1\xe8\x03\x8c\xf1\x9f\x03\x00\xb8\xcf\x1f\x8b\x00\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\x01@\x89\x01\x00%\x06\x00%\x06\x00\x94\x18\x00\x94\x18\x00\x94\x18\x00Pb\x00Pb\x00@\x89\xd9xu\xa99\x9f\xda\xb3^\xbb\x95a\xea\x9a\xe3\xbd@\x89A\x89\x8d\xfe\xf7\xaf9\xde\x0b\x94\x18\x94\xd8\xe8\x7f\xf3\x9a\xe3\xbd@\x89A\x89\x8d\xfe7\xaf9\xde\x0b\x94\x18f

'The Grand Total, which is the balance amount before payment, is **376.40**.'

In [17]:
system_prompt = """
               You are a specialist in comprehending receipts.
               Input images in the form of receipts will be provided to you,
               and your task is to respond to questions based on the content of the input image.
               """
#system_prompt = "Convert Invoice data into json format with appropriate json tags as required for the data in image "
# image_path = "/content/handwritten.png"

user_prompt = "Convert Invoice data into json format with appropriate json tags as required for the data in image "
     

In [18]:
output = gemini_output(image_path, system_prompt, user_prompt)
output

'```json\n{\n  "receipt_details": {\n    "business_info": {\n      "name": "The Lone Pine",\n      "address": "43 Manchester Road, 12480 Brisbane, Australia",\n      "phone": "617-3236-6207"\n    },\n    "transaction_info": {\n      "invoice_number": "08000008",\n      "table_number": "25",\n      "date": "09/04/08",\n      "time": "12:45"\n    },\n    "items": [\n      {\n        "item": "Carlsberg Bottle",\n        "quantity": 2,\n        "price": 16.00\n      },\n      {\n        "item": "Heineken Draft Standard",\n        "quantity": 3,\n        "price": 24.60\n      },\n      {\n        "item": "Heineken Draft Half Liter",\n        "quantity": 1,\n        "price": 15.20\n      },\n      {\n        "item": "Carlsberg Bucket (5 bottles)",\n        "quantity": 2,\n        "price": 80.00\n      },\n      {\n        "item": "Grilled Chicken Breast",\n        "quantity": 4,\n        "price": 74.00\n      },\n      {\n        "item": "Sirloin Steak",\n        "quantity": 3,\n        "pri

In [19]:
from IPython.display import Markdown
Markdown(output)

```json
{
  "receipt_details": {
    "business_info": {
      "name": "The Lone Pine",
      "address": "43 Manchester Road, 12480 Brisbane, Australia",
      "phone": "617-3236-6207"
    },
    "transaction_info": {
      "invoice_number": "08000008",
      "table_number": "25",
      "date": "09/04/08",
      "time": "12:45"
    },
    "items": [
      {
        "item": "Carlsberg Bottle",
        "quantity": 2,
        "price": 16.00
      },
      {
        "item": "Heineken Draft Standard",
        "quantity": 3,
        "price": 24.60
      },
      {
        "item": "Heineken Draft Half Liter",
        "quantity": 1,
        "price": 15.20
      },
      {
        "item": "Carlsberg Bucket (5 bottles)",
        "quantity": 2,
        "price": 80.00
      },
      {
        "item": "Grilled Chicken Breast",
        "quantity": 4,
        "price": 74.00
      },
      {
        "item": "Sirloin Steak",
        "quantity": 3,
        "price": 96.00
      },
      {
        "item": "Coke",
        "quantity": 1,
        "price": 3.50
      },
      {
        "item": "Ice Cream",
        "quantity": 5,
        "price": 18.00
      }
    ],
    "summary": {
      "subtotal": 327.30,
      "sales_gov_tax": {
        "percentage": "5%",
        "amount": 16.36
      },
      "service_charge": {
        "percentage": "10%",
        "amount": 32.73
      },
      "grand_total": 376.40
    },
    "payment_info": {
      "method": "Cash",
      "amount_paid": 400.00,
      "change": 23.60
    },
    "footer_messages": [
      "Thank you and see you again!",
      "John",
      "Bring this bill back within the next 10 days and get 15% discount on that day's food bill..."
    ]
  }
}
```

---

In [ ]:
system_prompt = """
               You are a specialist in comprehending receipts.
               Input images in the form of receipts will be provided to you,
               and your task is to respond to questions based on the content of the input image.
               """
#system_prompt = "Convert Invoice data into json format with appropriate json tags as required for the data in image "
image_path = "/home/manish/Desktop/projects/etfp/temp/image_19.png"

user_prompt = "Convert Invoice data into json format with appropriate json tags as required for the data in image "

output = gemini_output(image_path, system_prompt, user_prompt)
     

from IPython.display import Markdown
output = Markdown(output)
output

```json
{
  "invoice_header": {
    "invoice_no": "12",
    "dated": "6-May-2021",
    "page": "3"
  },
  "biller_details": {
    "name": "High Furnace Manufacturer",
    "address": "Unit 1, Sector 2, Khanpur road Jaisalmer, Rajasthan.",
    "gst": "23AM2359012D1ZW"
  },
  "consignee_details": {
    "name": "JK Energies Limited",
    "address": "House No. 1, Kolar road, Rajgarh, Madhya Pradesh.",
    "gst": "23KM4589012B1ZS"
  },
  "shipping_delivery_details": {
    "delivery_note": null,
    "mode_terms_of_payment": null,
    "supplier_ref": "12",
    "other_reference": null,
    "buyer_order_no": null,
    "buyer_order_dated": null,
    "despatch_document_no": null,
    "delivery_note_date": null,
    "despatched_through": null,
    "destination": "RAJGARH",
    "bill_of_lading_rr_no": null,
    "motor_vehicle_no": null,
    "terms_of_delivery": null
  },
  "item_summary": [
    {
      "description": "SGST R/OFF",
      "amount": "39,521.30"
    },
    {
      "description": "Less",
      "amount": "-0.03"
    }
  ],
  "financial_summary": {
    "total_amount_chargeable": {
      "value": "10,43,481.00",
      "in_words": "INR Ten Lakh Forty Three Thousand Four Hundred Eighty One Only"
    },
    "hsn_sac_breakdown": [
      {
        "hsn_sac": "85414012",
        "taxable_value": "4,86,641.28",
        "central_tax": {
          "rate": "2.50%",
          "amount": "12,166.02"
        },
        "state_tax": {
          "rate": "2.50%",
          "amount": "12,166.02"
        },
        "total_tax_amount": "24,332.04"
      },
      {
        "hsn_sac": "730890",
        "taxable_value": "1,62,304.77",
        "central_tax": {
          "rate": "9%",
          "amount": "14,607.43"
        },
        "state_tax": {
          "rate": "9%",
          "amount": "14,607.43"
        },
        "total_tax_amount": "29,214.86"
      },
      {
        "hsn_sac": "841391",
        "taxable_value": "74,777.69",
        "central_tax": {
          "rate": "9%",
          "amount": "6,729.99"
        },
        "state_tax": {
          "rate": "9%",
          "amount": "6,729.99"
        },
        "total_tax_amount": "13,459.98"
      },
      {
        "hsn_sac": "84137010",
        "taxable_value": "2,40,714.69",
        "central_tax": {
          "rate": "2.50%",
          "amount": "6,017.86"
        },
        "state_tax": {
          "rate": "2.50%",
          "amount": "6,017.86"
        },
        "total_tax_amount": "12,035.72"
      }
    ],
    "hsn_sac_breakdown_totals": {
      "total_taxable_value": "9,64,438.43",
      "total_central_tax_amount": "39,521.30",
      "total_state_tax_amount": "39,521.30",
      "total_tax_amount": "79,042.60"
    },
    "tax_amount_in_words": "INR Seventy Nine Thousand Forty Two and Sixty paise Only"
  },
  "company_bank_details": {
    "bank_name": "IDFC FIRST BANK",
    "account_no": null,
    "branch_ifs_code": null
  },
  "company_pan": null,
  "declaration": "We declare that this invoice shows the actual price of the goods described and that all particulars are true and correct.",
  "jurisdiction": "SUBJECT TO JAIPUR JURISDICTION",
  "note": "This is a Computer Generated Invoice"
}
```

---

In [1]:
import google.generativeai as genai 

google_api = 'AIzaSyAMlYqwvuQgekl8nlqc56XTqJVBufszrBU'
genai.configure(api_key=google_api)

# Model Configuration
MODEL_CONFIG = {
  "temperature": 0.2,
  "top_p": 1,
  "top_k": 32,
  "max_output_tokens": 4096,
}

## Safety Settings of Model
safety_settings = [
  {
    "category": "HARM_CATEGORY_HARASSMENT",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  },
  {
    "category": "HARM_CATEGORY_HATE_SPEECH",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  },
  {
    "category": "HARM_CATEGORY_SEXUALLY_EXPLICIT",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  },
  {
    "category": "HARM_CATEGORY_DANGEROUS_CONTENT",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  }
]

model = genai.GenerativeModel(
                            # model_name='gemini-2.5-flash',  # 6 min takes to 4 image 
                            model_name='gemini-2.5-pro',
                              generation_config=MODEL_CONFIG,
                              safety_settings=safety_settings)
model



def gemini_output(image_path,
                  system_prompt,
                  user_prompt):
    
    
    input_prompt = [system_prompt, image_path, user_prompt]
    response = model.generate_content(input_prompt)

    return response.text




/home/manish/anaconda3/envs/cuda121/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
def process_multiple_images(image_paths,
                            system_prompt,
                            user_prompt):

    all_results = []

    for image_path in image_paths:
        result = gemini_output(image_path, system_prompt, user_prompt)
        all_results.append({
            result
        })

    return all_results
    



In [3]:
system_prompt = """
               You are a specialist in comprehending receipts.
               Input images in the form of receipts will be provided to you,
               and your task is to respond to questions based on the content of the input image.
               """

image_path = "/home/manish/Desktop/projects/etfp/temp/image_19.png"


user_prompt = "Convert Invoice data into json format with appropriate json tags as required for the data in image "

output = process_multiple_images(image_path, system_prompt, user_prompt)
output


ValueError: Invalid operation: The `response.text` quick accessor requires the response to contain a valid `Part`, but none were returned. The candidate's [finish_reason](https://ai.google.dev/api/generate-content#finishreason) is 2.

[{'I understand you\'ve provided a file path `/home/manish/Desktop/projects/etfp/temp/image_19.png` for the receipt image. However, as an AI, I cannot directly access local file paths on your computer or browse your desktop.\n\nTo help you, please provide the content of the receipt image. You can do this by:\n\n1.  **Uploading the image directly** if the interface allows it.\n2.  **Describing the key information** from the invoice in detail (e.g., "Invoice number: INV-2023-001, Date: 2023-10-26, Vendor: ABC Store, Address: 123 Main St, Item 1: Coffee, Qty: 2, Price: 3.50, etc.").\n\nOnce I have the image content, I will convert the invoice data into the requested JSON format.'},
 {"I understand you'd like me to process the receipt image located at `/home/manish/Desktop/projects/etfp/temp/Bill Image_Receipt.png` and convert its data into JSON format.\n\nHowever, as an AI, I cannot directly access local files on your computer or view images from a file path like the one you provided. My capabilities are limited to processing text inputs.\n\nTo help you, please provide the text content of the receipt (e.g., by transcribing it, or describing the key details like store name, date, items, quantities, prices, subtotal, tax, total, etc.). Once you provide the text, I will gladly convert it into the requested JSON format."}]

----

In [13]:
from google import generativeai as genai 
from pydantic import BaseModel


class Recipe(BaseModel):
    recipe_name: str 
    ingredients: list[str]

# Recipe(recipe_name="chokha bhat", ingredients=["potato", "oil", "chilli", "salt", "rice", "water",])

genai.configure(api_key="AIzaSyAMlYqwvuQgekl8nlqc56XTqJVBufszrBU")
model = genai.GenerativeModel(model_name="gemini-2.5-pro") # gemini-2.5-pro, gemini-2.5-flash
output = model.generate_content("tell my who are you?")
output


response:
GenerateContentResponse(
    done=True,
    iterator=None,
    result=protos.GenerateContentResponse({
      "candidates": [
        {
          "content": {
            "parts": [
              {
                "text": "Of course. I am a large language model, trained by Google."
              }
            ],
            "role": "model"
          },
          "finish_reason": "STOP",
          "index": 0
        }
      ],
      "usage_metadata": {
        "prompt_token_count": 7,
        "candidates_token_count": 14,
        "total_token_count": 1108
      },
      "model_version": "gemini-2.5-pro"
    }),
)

In [5]:
from datetime import datetime
from typing import Optional
from pydantic import BaseModel

class User(BaseModel):
    id: int 
    name: str = 'john Doe'
    signup_ts: Optional[datetime] = None
    friends: list[int] = []


User(id=1)

User(id=1, name='john Doe', signup_ts=None, friends=[])

---